### PR TITLE
Fix coordinate transformations near Earth

### LINKED ISSUE #4221: Fix bug in GCRS<->ICRS conversion for near-earth objects
As has been discussed a fair amount in the astroplan project, as well as on the mailing list (the thread starting with https://mail.scipy.org/pipermail/astropy/2015-August/003907.html), there's a significant bug with the GCRS<->ICRS transformation for coordinates that are close to the earth (seems to be within <~1 AU).  For a regression test demonstrating the problem, see the first few commits in https://github.com/eteq/astropy/tree/fix-gcrs-trans 

I'm pretty sure I understand the source of this (basically, it's due to an incomplete workaround to some of ERFA's limitations), and hence am assigning it to myself.  Making this issue mainly so that we can try to get it fixed before a final v1.1 release.  

An item to consider, though: once fixed, should this be included in the 1.0.x branch?  While it is a bug in that the transforms aren't right, it is *technically* an API change in case someone worked around it already... (@astrofrog and/or @embray)?

### PR DESCRIPTION
**Pull request by eteq**
This PR closes #4221.    It turns out the problems there were a bit more complicated because there were two intertwined bugs:

* The way distances were being calculated for the ICRS<->CIRS transformations and the CIRS<->AltAz transformations was wrong in a somewhat subtle way.  This lead to errors ranging from a few thousand km to as much as an AU, which is primarily important inside the solar system.  Because the angular locations are computed using ERFA, they are not substantially affected, only distances.
* When doing the CIRS<->AltAz transformations, computing the distance change requires working with the `AltAz`'s `EarthLocation`.  It turns out that `ITRS` coordinate corresponding to the `EarthLocation` was *not* correctly being set to the `obstime` of the `AltAz`, but rather defaulting to J2000.  So distance were all computed relative to the J2000 location of the Earth rather than the current `obstime`.  So this lead to distance errors of order 1-2 AU at any time other than J2000.  This PR addresses that problem by creating a way to get the `ITRS` from an `EarthLocation` with a particular `obstime`, and then using that in the transformation.

There's actually a lot more in this PR then just those fixes, but it's essentially all tests (new ones or reorganization of old ones), because figuring out the root cause of these issues involved a ton of testing the various transform steps.  In any event, these tests should make solar-system scale transformations a fair bit better tested.

Note that I also tried backporting this to v1.0.x - there are some conflicts... nothing major, but I should probably do the backporting once this gets merged so that we can be sure the right things get backported.  Backporting to 1.1.x should be clean, though.

@embray @astrofrog - this fix should actually go into both 1.0.x *and* 1.1.x, so I'm correct that I should milestone it for 1.0.6, right?

cc @bmorris3 @adrn @mhvk

### CODE REVIEW

**Comment by bmorris3** [comment]
Background for those other than @eteq: this bug was discovered when I was trying to compute the position of the moon using `jplephem` and transforming cartesian coordinates to `AltAz`. 

[My test code](https://gist.github.com/bmorris3/a0bf3057ad078cc0616e#file-test-fix-gcrs-py) computes the alt/az position of the moon using PyEphem and compares it to the alt/az position I get from `jplephem` after various transformations with `SkyCoord`s. The test code runs on v1.0.5 and demonstrates the large difference between the answers (i.e., the bug):
```
PyEphem (Alt, Az): (55.4920153458 deg, 239.031895666 deg)
jplephem (Alt, Az): (34.177292986 deg, 135.374232837 deg)
```
When I run the same script with `astropy` running from this branch (@eteq:fix-gcrs-trans), I get `RuntimeError: maximum recursion depth exceeded in __instancecheck__`.

File: astropy/coordinates/earth.py [hunk 1]
```diff
@@ -368,13 +368,33 @@ def to_geocentric(self):
...
-        this object.
+        this object at the default ``obstime``.
         """
-        #potential circular imports prevent this from being up top
+        #circular imports prevent this from being up top
         from .builtin_frames import ITRS
 
         return ITRS(x=self.x, y=self.y, z=self.z)
 
+    def get_itrs(self, obstime):
```
Comment (mhvk) [inline astropy/coordinates/earth.py]: If we define `obstime=None` here (or whatever is necessary to make it use the "default obstime"), you can just do `itrs = property(get_itrs)` and remove the duplicate definition above.

File: astropy/coordinates/earth.py [hunk 2]
```diff
@@ -368,13 +368,33 @@ def to_geocentric(self):
...
+        obstime : `~astropy.time.Time`
+            The ``obstime`` to apply to the new `~astropy.coordinates.ITRS`
+
+        Returns
+        -------
+        itrs : `~astropy.coordinates.ITRS`
+            The new object in the ITRS frame
+        """
+        #circular imports prevent this from being up top
+        from .builtin_frames import ITRS
```
Comment (mhvk) [inline astropy/coordinates/earth.py]: Could the import be at the bottom?

**Comment by mhvk** [comment]
@eteq - so far only looked at the `EarthLocation` changes -- see the small inline comments. A larger one (probably for a separate issue if at all relevant): `EarthLocation` was always a bit of a hack to get sidereal times going. Should it just be a ITRS subclass?

File: astropy/coordinates/earth.py (hunk 2 above)
Comment (eteq) [inline astropy/coordinates/earth.py]: That alone didn't work, but it turns out an alternative solution was to move the import of `EarthLocation` in `baseframe` to the bottom.  Will push up a commit with that shortly.

File: astropy/coordinates/earth.py (hunk 1 above)
Comment (eteq) [inline astropy/coordinates/earth.py]: good point! will do.

**Comment by eteq** [comment]
@bmorris3 : Hmm, I got a few errors relating to `EarthLocation` (unrelated to this specific set of changes) when I first ran it, but after tweaking it a bit I got this:
```
With astropy v1.1.dev13915
PyEphem (Alt, Az): (55.4920153458 deg, 239.031895666 deg)
jplephem (Alt, Az): (57.0399296347 deg, 238.705846051 deg)
```
which is a lot closer than before.  Is there any chance you're running from an old version of my `fix-gcrs-trans` branch?  I was getting recursion errors a while back for some in-progress work, but those *should* have been fixed...  In the meantime, I'll try to take a closer look at your example code to try to figure out the origin of the remaining few degrees discrepancy.

**Comment by eteq** [comment]
@mhvk - You may well be right that `EarthLocation` should be re-implemented as ITRS... although there is one important distinction right now: ITRS has a specific obstime attached, while EarthLocation does not.  So EarthLocation is sort of a "location on the earth at *any* time", while `ITRS` is "location in space-time, but written down in a way that's somewhat more convenient for translating to specific parts of the Earth itself".  I think this distinction is useful, but it certainly might be worthwhile to take a look at how we might re-work `EarthLocation` into something that looks more like a frame class.

I think that's better tackled separately for v1.2, though, as it's definitely not just a "bug fix".

**Comment by mhvk** [comment]
@eteq - OK, new issue on EarthLocation possibly being an ITRS subclass: #4261

**Comment by eteq** [comment]
Good point @mhvk - just pushed up a new commit doing just what you suggested there.

**Comment by embray** [comment]
So is this ready to be merged?  I'd like to release v1.0.6 ASAP.

**Comment by eteq** [comment]
@embray Well, it doesn't look like anyone has reviewed the transformation code yet... So we *could* merge it given the tests are passing and such, but it would probably be better if someone other than me would review the distance calculations... (@adrn or @astrofrog?)

**Comment by mhvk** [comment]
@eteq - from https://github.com/astropy/astropy/pull/4254#issuecomment-150049124, it seemed there was still a problem at the ~degree level. Has that been resolved?

**Comment by embray** [comment]
It can be delayed for 1.0.7 if need be.  Or merged provisionally since it at least superficially fixes the problem...??

**Comment by adrn** [comment]
Sorry -- won't have time to review in detail before, say, Nov. 1 :(

**Comment by eteq** [comment]
Alright, I made some more adjustments to the test code and also tested this on a branch that includes both this (#4254) *and* #4255 - see  https://gist.github.com/eteq/08dba433082e57b7f2af for what actually yields this:
```
With astropy v1.1.dev13939
PyEphem (Alt, Az, Dist): (55.4920153458 deg, 239.031895666 deg, 356424848.107 m)
Skyfield (Alt, Az, Dist): (55.32651159 deg, 239.298837015 deg, 0.00238248020632 AU)
Skyfield(de430) (Alt, Az, Dist): (55.3265117716 deg, 239.298837058 deg, 0.00238248020483 AU)
jplephem/Astropy (Alt, Az, Dist): (56.0578669005 deg, 239.043478848 deg, 356423493.806 m)
jplephem/Astropy(ICRS) (Alt, Az, Dist): (56.0596259457 deg, 239.03677597 deg, 356421587.215 m)
```
Whereas a branch with just this gives the last two as:
```
jplephem/Astropy (Alt, Az, Dist): (57.0399296347 deg, 238.705846051 deg, 357213053.992 m)
jplephem/Astropy(ICRS) (Alt, Az, Dist): (56.0596259457 deg, 239.03677597 deg, 356421587.215 m)
```

So with that, the difference @mhvk is mentioning between PyEphem and astropy are at the <~half a degree level, which is *also* the disagreement between Skyfield and PyEphem (which are both by the same author and conceptually similar, but with somewhat different ephemerides).  At that level I'm willing to chalk it up to subtle differences in how the various frames are to be interpreted (e.g., the ICRS/GCRS difference in the last two lines are because ICRS->AltAz includes aberration, light deflection by the sun, etc., which of course is not relevant for the moon but might be for objects elsewhere in the solar system).

**Comment by eteq** [comment]
So unless someone thinks they can merge it within the next, say, day (is that your timescale for 1.0.6, @embray?), we should probably just merge and do a "live" review, as this is a big improvement over master/1.0.5 ...

Of course there's nothing to stop us from merging and then getting a more thorough review when someone has time, with a follow-on PR fixing any major issues.

**Comment by embray** [comment]
I'm thinking more like the next hour, but yeah :)

File: astropy/coordinates/builtin_frames/cirs_observed_transforms.py [hunk 3]
```diff
@@ -86,17 +92,32 @@ def altaz_to_cirs(altaz_coo, cirs_frame):
...
 
     if isinstance(altaz_coo.data, UnitSphericalRepresentation):
         distance = None
     else:
-        #now we get the distance as just the cartesian
-        #distance from the earth location to the coordinate location
-        distance = altaz_coo.separation_3d(altaz_coo.location.itrs)
+        locitrs = altaz_coo.location.get_itrs(altaz_coo.obstime)
+
+        # To compute the distance in a way that is reversable with cirs_to_altaz
```
Comment (eteq) [inline astropy/coordinates/builtin_frames/cirs_observed_transforms.py]: This block is in particular need of review/somebody else considering it.  The geometry is rather confusing, but most importantly, I'm not certain how numerically stable this solution to the SSA trig problem is. So if anyone else knows a trick for this particular geometric calculation I'd love to hear it.

**Comment by eteq** [comment]
Hah, ok, that works too @embray - if you want to pull the trigger to make sure this makes 1.0.6, feel free to do so.  I'd still appreciate whatever review this can get by someone willing to take the time to wrap their brain around the distance calculations, though!

**Comment by embray** [comment]
@eteq Maybe open an issue to ensure that this gets further review?

**Comment by mhvk** [comment]
Agreed that it is better to have something we at least *think* is correct! I'm happy to have this merged as long as we have a new issue with a very specific test case (the moon seems fine...). Ideally, for that, we also involve @brandon-rhodes.

**Comment by embray** [comment]
Just to note, when I backported this to v1.0.6 I had to remove some stuff related to galactic coordinates, I think it was?, that wasn't applicable to 1.0.x.  But I confirmed that the other tests passed.

**Comment by eteq** [comment]
@embray - am I right that 60edc624aa0cd2feec36368802f6189816166646 is the commit that backported this to v1.0.x ?  That seems to be the right diff, but the commit message doesn't seem right... That's not really a big deal, except that I wonder if it will confuse some of the backporting tools/log parsers?

But the backporting all looks find in terms of the actual content!  Will create a new issue shortly to prompt someone to do a review.

**Comment by embray** [comment]
Weird about the commit message.  I think I might know how that happened.  Now I need to make sure the other commit messages on that branch are correct....   I will fix it.

**Pull request merged**